modules/Bio/EnsEMBL/Utils/Tree/Interval/Immutable.pm

```diff
@@ -0,0 +1,320 @@
+
+=head1 METHODS
+
+=cut
+
+package Bio::EnsEMBL::Utils::Tree::Interval::Immutable;
+
+use strict;
+
+use Data::Dumper;
+
+use Bio::EnsEMBL::Utils::Scalar qw(assert_ref);
+use Bio::EnsEMBL::Utils::Exception qw(throw);
+use Bio::EnsEMBL::Utils::Tree::Interval::Immutable::Node;
+use Bio::EnsEMBL::Utils::Interval;
+
+=head2 new
+
+=cut
+
+sub new {
+  my $caller = shift;
+  my $class = ref($caller) || $caller;
+
+  my $intervals = shift;
+  $intervals && assert_ref($intervals, 'ARRAY');
+  
+  my $self = bless({}, $class);
+
+  $self->{top_node} = $self->_divide_intervals($intervals);
+  return $self;
+}
+
+sub root {
+  return shift->{top_node};
+}
+
+# sub search {
+#   my ($self, $start, $end) = @_;
+#   defined $start or throw 'Undefined point or interval';
+
+#   # interval search
+#   if (defined $end) {
+#     my $result = [];
+#     for my $i ($start .. $end + 1) {
+#       push @{$result}, @{$self->search($i)};
+#     }
+
+#     return sort_by_begin(uniq($result));
+#   } 
+
+#   # point search 
+#   return $self->_search($self->{top_node}, $start, []);
+# }
+
+# sub _search {
+#   my ($self, $node, $point, $result) = @_;
+
+#   # can be optimized, see https://en.wikipedia.org/wiki/Interval_tree
+#   for my $i (@{$node->scenter}) {
+#     push @{$result}, $i if $i->start <= $point and $i->end >= $point;
+#   }
+
+#   if ($point < $node->x_center and $node->left_child) {
+#     my $left_results = $self->_search($node->left_child, $point, []);
+#     push @{$result}, @{$left_results} if $left_results;
+#   }
+
+#   if ($point > $node->x_center and $node->right_child) {
+#     my $right_results = $self->_search($node->right_child, $point, []);
+#     push @{$result}, @{$right_results} if $right_results;
+#   } 
+
+#   return uniq($result);
+# }
+
+sub query {
+  my ($self, $start, $end) = @_;
+  
+  my $interval;
+  if (defined $start) {
+    $end = $start unless defined $end;
+    $interval = Bio::EnsEMBL::Utils::Interval->new($start, $end);
+  }
+  
+  return [] unless $interval;
+  return $self->_query_point($self->root, $interval->start, []) if $interval->is_point;
+
+  my $result = [];
+  return $result unless $self->root or $interval->is_empty;
+
+  my $node = $self->root;
+  while ($node) {
+    if ($interval->contains($node->x_center)) {
+      push @{$result}, @{$node->s_center_beg};
+      $self->_range_query_left($node->left, $interval, $result);
+      $self->_range_query_right($node->right, $interval, $result);
+      last;
+    }
+    if ($interval->is_left_of($node->x_center)) {
+      foreach my $s_beg (@{$node->s_center_beg}) {
+	last unless $interval->intersects($s_beg);
+	push @{$result}, $s_beg;
+      }
+      $node = $node->left;
+    } else {
+      foreach my $s_end (@{$node->s_center_end}) {
+	last unless $interval->intersects($s_end);
+	push @{$result}, $s_end;
+      }
+      $node = $node->right;
+    }
+  }
+
+  return sort_by_begin(uniq($result));
```

Comment:
sort_by_begin should be used as $self->sort_by_begin
this happens a few times, so change all cases

Reply:
Sorry I don't see why it's a problem this way. This is a utility function working on intervals which is placed in this module for convenience as it is not used anywhere else. It is not strictly part of the definition of the tree.

Reply:
Won't fix.

Reply:
shouldn't it at least be &sort_by_begin ?

Reply:
Both yours and mine are valid Perl ways to invoke a subroutine. What do drawbacks you see in the latter?

Reply:
it's just not as easy to identify what is happening, but you're right they are both valid and not really an issue.

Reply:
I see, thanks both of you for the review.